File: lua/collatz-conjecture/collatz-conjecture.lua
local number = {
    is_even = function(x)
        return x % 2 == 0
    end,
    is_postive = function(x)
        return x > 0
    end,
    is_integer = function(x)
        return math.floor(x) == x
    end
}

local collatz = {}

function collatz.number(x)
    if number.is_even(x) then
        return x / 2
    else
        return 3 * x + 1
    end
end

function collatz.steps(x, prev_steps)
    if x == 1 then
        return prev_steps
    else
        return collatz.steps(collatz.number(x), prev_steps + 1)
    end
end

return function(n)
    assert(
        number.is_postive(n) and number.is_integer(n),
        "Only positive numbers are allowed"
    )
    return collatz.steps(n, 0)
end

File: javascript/collatz-conjecture/collatz-conjecture.js
//
// This is only a SKELETON file for the 'Collatz Conjecture' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

//@ts-check

/**
 * 
 * @param {number} x 
 * @returns {boolean}
 */
const isPositive = ( x ) => x > 0;

/**
 * 
 * @param {number} x 
 * @returns {boolean}
 */
const isInteger = Number.isInteger;

/**
 * 
 * @param {number} x 
 * @returns {boolean}
 */
const isPositiveInteger = ( x ) => ( isPositive( x ) && isInteger( x ) );

/**
 * 
 * @param {number} x 
 * @returns {boolean}
 */
const isEven = ( x ) => (x % 2 == 0);

/**
 * 
 * @param {number} x 
 * @returns {number}
 */
const collatz = ( x ) => {
  
  if( isEven( x ) )
    return x / 2;
  
  else
    return 3 * x + 1

};

const calculateCollatzSteps = ( x ) => {

  const noStep = 0;
  const oneStep = 1;
  const collatzStop = 1;

  if( x == collatzStop )
    return noStep;

  return oneStep + calculateCollatzSteps( collatz( x ) );

}


/**
 * 
 * @param {number} x
 * @returns {number} 
 */
export const steps = ( x ) => {

  if( !isPositiveInteger( x ) )
    throw new Error( "Only positive numbers are allowed" );

  return calculateCollatzSteps( x );

};
